server/api/src/rsDataObjChksum.cpp

```diff
@@ -1,97 +1,400 @@
+        }
+
+        const auto client_set_replica_number = kvp.contains(REPL_NUM_KW);
+        const auto client_set_resource_name = kvp.contains(RESC_NAME_KW);
+
+        if (client_set_replica_number && client_set_resource_name) {
+            return USER_INCOMPATIBLE_PARAMS;
+        }
+
+        const auto client_targeted_specific_replica = (client_set_replica_number || client_set_resource_name);
+        const auto client_set_all_flag = kvp.contains(CHKSUM_ALL_KW);
+
+        if (client_set_all_flag && client_targeted_specific_replica) {
+            return USER_INCOMPATIBLE_PARAMS;
+        }
+
+        // Verify that the client is allowed to use the administrative flag.
+        if (kvp.contains(ADMIN_KW) && !irods::is_privileged_client(_comm)) {
+            return CAT_INSUFFICIENT_PRIVILEGE_LEVEL;
+        }
+
+        if (const auto ec = getDataObjInfoIncSpecColl(&_comm, &_dataObjInp, &_replicas); ec < 0) {
+            return ec;
+        }
+
+        if (client_targeted_specific_replica) {
+            // The client is allowed to verify checksum information for a single stale replica.
+            // Targeting a specific replica supports verifying the following:
+            // - The existence of a checksum for that replica.
+            // - The physical size of the replicas vs what is recorded in the catalog.
+            if (const auto ec = sortObjInfoForOpen(&_replicas, &_dataObjInp.condInput, /* write flag */ 0); ec < 0) {
+                return ec;
+            }
+
+            if (const auto ec = verifyVaultSizeEqualsDatabaseSize(_comm, *_replicas); ec < 0) {
+                std::string msg;
+
+                if (ec == USER_FILE_SIZE_MISMATCH) {
+                    msg = fmt::format("ERROR: Physical size does not match size in catalog for replica [{}].", _replicas->replNum);
+                }
+                else {
+                    msg = fmt::format("ERROR: Could not verify size for replica [{}].", _replicas->replNum);
+                }
+
+                addRErrorMsg(&_comm.rError, ec, msg.data());
+            }
+
+            if (std::strlen(_replicas->chksum) == 0) {
+                const auto msg = fmt::format("WARNING: No checksum available for replica [{}].", _replicas->replNum);
+                addRErrorMsg(&_comm.rError, CAT_NO_CHECKSUM_FOR_REPLICA, msg.data());
+            }
+
+            return 0;
+        }
+
+        auto good_replicas = get_good_replicas(_replicas);
+
+        report_number_of_stale_replicas_skipped(_replicas, _comm.rError);
+        report_replicas_with_mismatch_size_info(_comm, good_replicas);
+
+        if (report_replicas_without_checksums(good_replicas, _comm.rError)) {
+            // Return immediately because this function requires that all replicas
+            // have a checksum after this point.
+            return 0;
+        }
+
+        //
+        // At this point, all replicas are guaranteed to have a checksum.
+        //
+
```

Comment:
consider possibility of other replicas being created in the meantime?    or are we within a giant transaction?

Reply:
this function uses only the list of replicas retrieved by `getDataObjInfoIncSpecColl`, so new replicas are not a concern at this point.

and no, this code is not within transaction.